File: app/analytics/smoothing.py
from __future__ import annotations

from collections import deque
from enum import Enum


class SmoothingMode(str, Enum):
    NONE = "NONE"
    EMA = "EMA"
    MOVING_AVG = "MOVING_AVG"


class MultiChannelSmoother:
    """Applies smoothing independently per finger channel."""

    def __init__(self, mode: SmoothingMode = SmoothingMode.EMA, ema_alpha: float = 0.25, window: int = 5):
        self.mode = mode
        self.ema_alpha = max(0.01, min(0.99, float(ema_alpha)))
        self.window = max(1, int(window))

        self._ema_state: list[float | None] = [None, None, None, None]
        self._ma_buf = [deque(maxlen=self.window) for _ in range(4)]

    def reset(self) -> None:
        self._ema_state = [None, None, None, None]
        self._ma_buf = [deque(maxlen=self.window) for _ in range(4)]

    def reconfigure(self, mode: SmoothingMode, ema_alpha: float, window: int) -> None:
        self.mode = mode
        self.ema_alpha = max(0.01, min(0.99, float(ema_alpha)))
        self.window = max(1, int(window))
        self.reset()

    def apply(self, values: tuple[float, float, float, float]) -> tuple[float, float, float, float]:
        if self.mode == SmoothingMode.NONE:
            return values

        if self.mode == SmoothingMode.EMA:
            out: list[float] = []
            for i, value in enumerate(values):
                prev = self._ema_state[i]
                if prev is None:
                    self._ema_state[i] = float(value)
                else:
                    self._ema_state[i] = (self.ema_alpha * float(value)) + ((1.0 - self.ema_alpha) * prev)
                out.append(self._ema_state[i] if self._ema_state[i] is not None else float(value))
            return tuple(out)  # type: ignore[return-value]

        # Moving average
        out_ma: list[float] = []
        for i, value in enumerate(values):
            self._ma_buf[i].append(float(value))
            out_ma.append(sum(self._ma_buf[i]) / max(1, len(self._ma_buf[i])))
        return tuple(out_ma)  # type: ignore[return-value]

File: packages/core/src/smoothing.ts
import type { Finger4 } from './types.ts';
import type { SmoothingMode } from './settings.ts';

export class MultiChannelSmoother {
  private mode: SmoothingMode;
  private emaAlpha: number;
  private window: number;
  private emaState: Array<number | null> = [null, null, null, null];
  private maBuf: number[][] = [[], [], [], []];

  constructor(mode: SmoothingMode = 'EMA', emaAlpha = 0.25, window = 5) {
    this.mode = mode;
    this.emaAlpha = Math.max(0.01, Math.min(0.99, emaAlpha));
    this.window = Math.max(1, Math.floor(window));
  }

  reset(): void {
    this.emaState = [null, null, null, null];
    this.maBuf = [[], [], [], []];
  }

  reconfigure(mode: SmoothingMode, emaAlpha: number, window: number): void {
    this.mode = mode;
    this.emaAlpha = Math.max(0.01, Math.min(0.99, emaAlpha));
    this.window = Math.max(1, Math.floor(window));
    this.reset();
  }

  apply(values: Finger4): Finger4 {
    if (this.mode === 'NONE') return values;

    if (this.mode === 'EMA') {
      const out: number[] = [];
      for (let i = 0; i < 4; i++) {
        const prev = this.emaState[i];
        if (prev === null) {
          this.emaState[i] = values[i];
        } else {
          this.emaState[i] = this.emaAlpha * values[i] + (1 - this.emaAlpha) * prev;
        }
        out.push(this.emaState[i] ?? values[i]);
      }
      return [out[0], out[1], out[2], out[3]];
    }

    const out: number[] = [];
    for (let i = 0; i < 4; i++) {
      this.maBuf[i].push(values[i]);
      if (this.maBuf[i].length > this.window) {
        this.maBuf[i].shift();
      }
      let sum = 0;
      for (const value of this.maBuf[i]) {
        sum += value;
      }
      out.push(sum / this.maBuf[i].length);
    }
    return [out[0], out[1], out[2], out[3]];
  }
}
